### DISCUSSION TITLE
I want to find the number of substructures.

### DISCUSSION THREAD

**Discussion by okikosuke** [rdkit/rdkit#5456] [Q&A]
I want to count the number of substructures in a compound, but I am having trouble. Please advise me.

Consider the following compounds with 0, 1, and 2 carboxyl groups in the compound.

```python
from rdkit import Chem

m0 = "CC(C1=CC=C(C(OC)=O)C=C1)=O"
m1 = "OC(C1=CC=C(C(OC)=O)C=C1)=O"
m2 = "OC(C1=CC=C(C(OC)=O)C=C1)=O"
```

````python
Chem.MolFromSmiles(m0)
````
![image](https://user-images.githubusercontent.com/74843794/180675648-c41164de-79b2-4aed-bb18-23204a3a6eb7.png)

````python
Chem.MolFromSmiles(m1)
````
![image](https://user-images.githubusercontent.com/74843794/180675669-cc51c902-1163-4539-b208-97522cb8c1cd.png)

````python
Chem.MolFromSmiles(m2)
````
![image](https://user-images.githubusercontent.com/74843794/180675689-7b407365-5d17-4802-b5cb-87bb0ad62f7b.png)

We want to count the number of carboxyl groups in m0, m1, and m2.
First, we store the compounds to be considered in a data frame.

```python
import pandas as pd

df = pd.DataFrame(
    data={"compound": ["m0", "m1", "m2"], 
          "smiles": [m0, m1, m2]}
)
df
```

![image](https://user-images.githubusercontent.com/74843794/180675718-c2bcfdc0-0eae-4ad4-b191-9f691c10f873.png)

Convert SMILES to Mol and keep them in a list.

```python
mol_list = [Chem.MolFromSmiles(x) for x in list(df["smiles"]) if x is not None]
```

Define the carboxyl group.

```python
cooh = Chem.MolFromSmiles("C(=O)O")
cooh
```

![image](https://user-images.githubusercontent.com/74843794/180675748-3da7adbb-d96d-4c0b-a9d4-0fef4002c7c6.png)

Count how many carboxyl groups are in m0.

```python
mol_list[0].HasSubstructMatch(cooh)
```

```
True
```

The number of carboxyl groups in m0 is 0, but the output is True.
Let's change the notation of SMILES.

```python
cooh_wc = Chem.MolFromSmiles("*C(=O)O")
cooh_wc
```

![image](https://user-images.githubusercontent.com/74843794/180675785-d80859e8-1b6d-49dc-a004-2a76a0b152e1.png)

```python
mol_list[0].HasSubstructMatch(cooh_wc)
```

```
False
```

```python
mol_list[1].HasSubstructMatch(cooh_wc)
```

```
False
```

The output is False even though there is one carboxyl group in m1.
Next, we search in SMARTS format.

```python
cooh_smarts = Chem.MolFromSmarts("C(=O)O")
cooh_smarts
```

![image](https://user-images.githubusercontent.com/74843794/180675846-273579d2-d7e6-4cff-bc64-a96eb0aeeac5.png)

```python
mol_list[0].HasSubstructMatch(cooh_smarts)
```

```
True
```

The number of carboxyl groups in m0 is 0, but the output is True.
I have tried adding hydrogen to the molecule, but it did not work. How can I solve this problem?

Translated with www.DeepL.com/Translator (free version)

**Comment by greglandrum** [comment]
If you want to get the number of substructure matches you should call `mol.GetSubstructMatches()`. That returns a tuple with all of the matches and you can just get its length:
```
In [2]: m = Chem.MolFromSmiles("OC(=O)CCC(O)=O")

In [3]: q = Chem.MolFromSmarts("*C(=O)O")

In [4]: len(m.GetSubstructMatches(q))
Out[4]: 2
```

As an aside: the `*` in `Chem.MolFromSmiles('*C(=O)O')` creates an atom with atomic number 0 which can only match other atoms with atom number 0. If you want the `*` to be a query then you need to either use `Chem.MolFromSmarts()` (`*` is a query which matches any atom in SMARTS) or use `Chem.AdjustQueryProperties()`. Here's an example of that:
```
n [5]: q2 = Chem.MolFromSmiles("*C(=O)O")

In [6]: len(m.GetSubstructMatches(q2))
Out[6]: 0

In [8]: ps = Chem.AdjustQueryParameters.NoAdjustments()

In [9]: ps.makeDummiesQueries = True

In [10]: q2a = Chem.AdjustQueryProperties(q2, ps)

In [11]: len(m.GetSubstructMatches(q2a))
Out[11]: 2
```

**Comment by okikosuke** [comment]
@greglandrum 

Thank you for your reply. However, this method does not achieve what I want. I want to count the number of carboxyl groups. In other words, I want to count "-COOH" but not "-COO-".

The method you have given me is as follows.

```python
from rdkit import Chem

m0 = Chem.MolFromSmiles("CC(C1=CC=C(C(OC)=O)C=C1)=O")
m0
```
![image](https://user-images.githubusercontent.com/74843794/180706244-b5dd2a1c-1c8e-4562-859d-9f5f7d7bccef.png)

m0 contains "-COO-" but no carboxyl group.

```python
q = Chem.MolFromSmarts("*C(=O)O")
q
```

![image](https://user-images.githubusercontent.com/74843794/180706436-70d429d7-feac-4863-bf2a-c4e2e55b855f.png)

```python
len(m0.GetSubstructMatches(q))
```

```
1
```

The output is 1, but m0 does not contain a carboxyl group, so we want it to be output as 0. What should I do?

**Reply by greglandrum** [reply]
I was just demonstrating how to count the number of times a particular substructure matches.
Getting the query itself correct is a different topic and requires some thought and experimentation.
It's best done using SMARTS.

A SMARTS which will only match carboxyl, but not carboxylate groups is: `*C(=O)-[OH1]`

Here's the RDKit's SMARTS reference: https://www.rdkit.org/docs/RDKit_Book.html#smarts-reference

**Comment by DavidACosgrove** [comment]
I suggest you read some documentation on how SMARTS works.  What you want
to do is very easy with SMARTS.  Try
https://www.daylight.com/dayhtml_tutorials/languages/smarts/, for example.



On Mon, Jul 25, 2022 at 6:46 AM okikosuke ***@***.***> wrote:

> @greglandrum <https://github.com/greglandrum>
>
> Thank you for your reply. However, this method does not achieve what I
> want. I want to count the number of carboxyl groups. In other words, I want
> to count "-COOH" but not "-COO-".
>
> The method you have given me is as follows.
>
> from rdkit import Chem
> m0 = Chem.MolFromSmiles("CC(C1=CC=C(C(OC)=O)C=C1)=O")m0
>
> [image: image]
> <https://user-images.githubusercontent.com/74843794/180706244-b5dd2a1c-1c8e-4562-859d-9f5f7d7bccef.png>
>
> m0 contains "-COO-" but no carboxyl group.
>
> q = Chem.MolFromSmarts("*C(=O)O")q
>
> [image: image]
> <https://user-images.githubusercontent.com/74843794/180706436-70d429d7-feac-4863-bf2a-c4e2e55b855f.png>
>
> len(m0.GetSubstructMatches(q))
>
> 1
>
> The output is 1, but m0 does not contain a carboxyl group, so we want it
> to be output as 0. What should I do?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rdkit/rdkit/discussions/5456#discussioncomment-3219045>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FUXEKN26ZZBD2EIMLTVVYS43ANCNFSM54QSEYMQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

**Comment by okikosuke** [comment]
@greglandrum @DavidACosgrove 
Thanks for the reply. Your comments have helped me solve the problem. I will refer to the site you gave me for the rules of SMARTS. For example, "*C(=O)-[OH1]" for carboxyl group, if you know of a site with specific examples, please let me know.

**Comment by okikosuke** [comment]
I have one more question.Suppose I have the following SMILES compounds.

```python
from rdkit import Chem

m ="O=C(OO1)C2=CC=CC=C2C1=O"
Chem.MolFromSmiles(m)
```

![image](https://user-images.githubusercontent.com/74843794/180944335-5d3294c6-2cbb-454f-bbcb-f7333f9441c0.png)

For this compound, what SMARTS would you recognize as the substructure of the ester bond (-C(=O)O-) and the bond with two oxygen atoms joined by a single bond (-O-O-)? I looked at the reference page but could not figure out how to specify it.
